packages/hierarchy-builder/src/hierarchy-builder/internal/TreeNodesReader.ts

```diff
@@ -26,17 +25,17 @@ export class TreeQueryResultsReader {
     // istanbul ignore next
     this._props = {
       parser: props?.parser ?? defaultNodesParser,
-      limit: props?.limit ?? DEFAULT_ROWS_LIMIT,
     };
   }
 
-  public async read(executor: IECSqlQueryExecutor, query: Omit<ECSqlQueryDef, "ctes">): Promise<ParsedHierarchyNode[]> {
+  public async read(executor: IECSqlQueryExecutor, query: Omit<ECSqlQueryDef, "ctes">, limit?: number): Promise<ParsedHierarchyNode[]> {
+    const nodeLimit = limit ?? DEFAULT_ROWS_LIMIT;
```

Comment:
We apply the limit on the query in `applyLimit` which is called by `HierarchyProvider` before sending the result here. Instead, I think `HierarchyProvider` should send the full `ECSqlQueryDef` here and `applyLimit` should be called from there too.

Reply:
Moved `applyLimit` call to `read` function.

Reply:
A couple of problems with the change:
- At the moment `applyLimit` is responsible for serializing `query.ctes` into the resulting ECSQL. However, the `query` argument of this function omits `ctes`, so I assume they never get included into the ECSQL?
- Mutating input argument is not a great idea (`query.ecsql = applyLimit(...)`), as that may cause unexpected consequences to the callers.

My suggestions:
1. Change `applyLimit` to only take ECSQL without CTEs. Leave serializing CTEs into the ECSQL for the caller.
2. In the `read` function:
  - Change `query` argument to not omit `ctes` attribute.
  - Serialize `ctes` and `ecsql` attributes of the given query into a local `ecsql` variable.
  - Pass the local `ecsql` to `applyLimit` to get the limit applied.

Reply:
Changed to your suggestions with a small difference:
Your suggestion: Serialize `ctes` and `ecsql` attributes of the given query into a local `ecsql` variable -> pass it to `applyLimit`.
My change: Pass provided `ecsql` variable to `applyLimit` -> combine serialized `ctes` and string from `applyLimit` and pass the value to `queryReader`
Made a different change than your suggestion because:
applyLimit currently wraps the ecsql inside another select clause:
![image](https://github.com/iTwin/presentation/assets/100586436/e3099f31-aa30-4caf-88ec-9c428248db7e)
And doing this with ctes added to the ecsql doesn't work. If you want your suggestion to work, then applyLimit shouldn't wrap the query inside another select clause, and should be implemented like: 
![image](https://github.com/iTwin/presentation/assets/100586436/96458a1a-db18-4913-8e41-cb41e9dae8f7)
Which approach should I use?